### ISSUE TITLE
Gunicorn consuming excessive amounts of memory

### ISSUE THREAD

**Issue by dralley** [pulp/pulpcore#2250]
**Version**
3.16.z

**Describe the bug**
Gunicorn consuming excessive amounts of memory, 3.5-4gb

**To Reproduce**
Unclear

**Expected behavior**
Probably not to have a single gunicorn process use 4gb of memory

**Additional context**

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=2035873
Katello forum discussion: https://community.theforeman.org/t/katello-4-5-foreman-3-3-memory-leak-in-gunicorn/29658/22

**Comment by bmbouter** [comment]
One workaround is to enable process recycling for gunicorn workers. This doesn't get at the root cause though, which I'm curious about. Also how does it grow over time, does it grow slowly or are there specific requests that leave large amounts of memory pinned?

**Comment by ipanova** [comment]
copy pasting for transparency so it's in one place
```
 To enable gunicorn worker recycling the procedure goes like this:

1.  Create a gunicorn.conf.py under /var/lib/pulp containing max_requests and max_requests_jitter values

# cat /var/lib/pulp/gunicorn.conf.py
max_requests = 1000
max_requests_jitter = 50

2.  restart the specific service systemctl restart pulpcore-api

3.  gunicorn workers will then be automatically recycled every 1000 requests (with a jitter to avoid all workers restarting at the same time) One can reduce the max_requests value to have more aggressive recycling.

```

**Comment by bmbouter** [comment]
The installer [passes the gunicorn options along via command line params](https://github.com/pulp/pulp_installer/blob/669dbdde710762c59a6840c1b97cca119d6f6c0a/roles/pulp_content/templates/pulpcore-content.service.j2#L27-L32), so we could use `--max-requests` and `--max-requests-jitter` options like we do today.

To implement this as a default I expected an installer + operator + single container to all make the changes, and pulpcore to document it's benefit.

**Comment by dralley** [comment]
This might not (fully) be a leak, look at what it's doing, it's fetching the full details of RPM packages from the /pulp/api/v3/rpm/packages/ endpoint.  So: serializing 100-1000 packages at a time.  That might still be a stretch but maybe if it's doing so really inefficiently?

**Comment by bmbouter** [comment]
@dralley I agree. My concern (besides the efficiency of that code) is that for subsequent requests the memory isn't released until the recycling occurs.

**Comment by dralley** [comment]
Added a link to a Katello forum discussion that is ongoing.  Some discoveries so far:  

* it's definitely the API server rather than anything else
* process recycling does help, but the memory is re-leaked extremely quickly
* it's worse with multiple API server processes running

**Comment by dralley** [comment]
Trimming the (RPM) changelogs appears to help.  So there is likely overlap with https://github.com/pulp/pulpcore/issues/2138

Sounds like it's responsible for not just memory usage issues but also potentially timeouts (there is a BZ for that as well)

**Comment by dralley** [comment]
Closing in favor of https://github.com/pulp/pulpcore/issues/2138

Katello / Satellite are the main impacted users right now, and they now have multiple workarounds.

**Comment by maartenbeeckmans** [comment]
Could this issue be reopened? This is not solved yet. We are running pulpcore version 3.21.4

We have some scripts in place for automating some pulp settings. The scripts that we are using use the python pulp client bindings, in these it is not possible to set the fields that the api returns in these bindings.

I can replicate the issue with the following api commands:

Ram usage (of the virtual machine) before the command: 3.7Gi of 24 Gi

```
# pulp rpm content list --name 'self-build-package' --limit 1000 | jq .[].location_href | wc -l
267
```

Ram usage (of the virtual machine) after execurting the command: 6.6Gi of 24Gi

Our pulp virtual machine ran out of memory once when the pulpcore-api service was using 12.7Gi of memory. This was cleared after restarting the service, but that should not be the case.

**Comment by dralley** [comment]
@maartenbeeckmans How old is this instance, and are you working with RHEL RPMs synced from at least 6 months ago?

**Comment by maartenbeeckmans** [comment]
Versions can be seen below, we are using the rpm's provided by theforeman.

```sh
# pulp status | jq .versions
[
  {
    "component": "core",
    "version": "3.21.4",
    "package": "pulpcore"
  },
  {
    "component": "rpm",
    "version": "3.18.9",
    "package": "pulp-rpm"
  },
  {
    "component": "file",
    "version": "1.11.1",
    "package": "pulp-file"
  },
  {
    "component": "deb",
    "version": "2.20.0",
    "package": "pulp_deb"
  },
  {
    "component": "container",
    "version": "2.14.3",
    "package": "pulp-container"
  }
]
```

**Comment by dralley** [comment]
I've tried a number of different debugging approaches, I know that the issue is all of the RPM changelogs loaded while listing the packages, but I don't know why they're leaking.  My best guess is that maybe the django queryset cache is being used on the viewset, but I'm not confident of that.

https://github.com/pulp/pulpcore/issues/2138 would help to work around it as does process recycling

**Comment by wbclark** [comment]
Hey @dralley , it appears the caching implemented in https://github.com/pulp/pulpcore/pull/2826 wasn't present in [Pulpcore 3.18](https://github.com/pulp/pulpcore/blob/3.18/pulpcore/app/access_policy.py) and first appears in [Pulpcore 3.20](https://github.com/pulp/pulpcore/blob/3.20/pulpcore/app/access_policy.py#L15).

I'm mentioning this because we (Satellite, in this case) received a hotfix request for https://github.com/pulp/pulpcore/pull/4090 and I'm creating a new BZ to track delivery of that fix (the existing BZ was already marked CLOSED ERRATA for with changes delivered in 6.13.2 so it's unsuitable to reuse it for this additional patch).

Do you need a new/separate upstream issue to track it as well? (I don't want to cause any problems for your tracking or automation if I attach this existing issue to an additional new downstream BZ... let me know if it's OK)

Thanks!

**Comment by dralley** [comment]
I don't think we need an additional upstream issue, at least not without something concrete
